### ISSUE TITLE
The data format of motion = gp.predict_gestures(args.audio,args.text)?

### ISSUE THREAD

**Issue by moonryul** [Svito-zar/gesticulator#64]
Hi, I enjoy your gesticulator. 
I would like to call the motion generator from Unity C# script, using pythonnet package.

I was experimenting how to get the motion from the gesticulator in C# script and apply it to an Unity humanoid character.
All things got done except that the motion displayed in Unity does not look right; The character remains in T pose. 

For testing I used  demo.py where we have the following code:

model = GesticulatorModel.load_from_checkpoint(  # MJ: model should be obtained in c# script once for all, not for every utterance
        args.model_file, inference_mode=True)
    # This interface is a wrapper around the model for predicting new gestures conveniently
    gp = GesturePredictor(model, feature_type)

    # 2. Predict the gestures with the loaded model
    motion = gp.predict_gestures(args.audio, args.text)
    
    => 
    
    The shape of motion is: (1, 528, 45). I thought that 45 refers to the number of Euler angles in the 15 selected joints
    specified in JointSelector(['Spine','Spine1','Spine2','Spine3','Neck','Neck1','Head','RightShoulder', 'RightArm', 'RightForeArm', 'RightHand', 'LeftShoulder', 'LeftArm', 'LeftForeArm', 'LeftHand']
    
    in
    
    data_pipe = Pipeline([
       ('dwnsampl', DownSampler(tgt_fps=fps,  keep_all=False)),
       ('root', RootTransformer('hip_centric')),
       ('mir', Mirror(axis='X', append=True)), ## MJ:The mirroring along the x axis is used as data augmentation technique;
       #  15 joint angles are selected:
       ('jtsel', JointSelector(['Spine','Spine1','Spine2','Spine3','Neck','Neck1','Head','RightShoulder', 'RightArm', 'RightForeArm', 'RightHand', 'LeftShoulder', 'LeftArm', 'LeftForeArm', 'LeftHand'], include_root=True)),
       ('exp', MocapParameterizer('expmap')), # Euler Angles to Exponential Map
       ('cnst', ConstantsRemover()),
       ('np', Numpyfier())
    ])
    
    from bvh2features.py
    
    Q1: Does not the motion include the rotation for the Hip joint? Did you assume that the rotation of the Hip joint is 
         constant (0,0,0) in Euler angles? I am confused because you mention "14 joints excluding the Hips" somewhere in your
        gesticulator paper.
         
    Q2:  The Euler angles in motion are in degrees rather than radians?    
    
    Thanks.

**Comment by moonryul** [comment]
Hi, after I send the above message, I found out that when you visualize the motion from the gesticulator by using
 visualize(motion.detach(), "temp.bvh", "temp.npy", "temp.mp4",
               start_t=0, end_t=motion_length_sec,
               data_pipe_dir='../gesticulator/utils/data_pipe.sav'),
               
               visualize() calls 
                write_bvh((data_pipe_dir,), # write_bvh expects a tuple
                motion.detach(),
              bvh_file,
              20)  ,

             which invokes inv_data = data_pipeline.inverse_transform(motion).
             
             I noticed that this inverse_transform converts exponential map to Euler angles. 
             It suggests that the motion from the gesticulator is in the format of exponential map.
            Is it right? 
            
            
           Then, I wonder how the shape of motion could be (1,528,45) for the 15 joints.

**Comment by Svito-zar** [comment]
Yes, Gesticulator spits our data in exponential map format which is then transformed into Euler angles

The shape of motion (1,528,45) means 528 frames of 3 joint rotations for 15 joints

**Comment by Svito-zar** [comment]
Btw, we have previously applied Gesticulator to an agent in Unity.
Here is the GitHub repository of that project: https://github.com/nagyrajmund/gesturebot

**Comment by moonryul** [comment]
2023년 6월 8일 (목) 오후 6:42, Taras Kucherenko ***@***.***>님이 작성:

> Yes, Gesticulator spits our data in exponential map format which is then
> transformed into Euler angles
>
> The shape of motion (1,528,45) means 528 frames of 3 joint rotations for
> 15 joints
>

=>>Thanks for the quick response.
So the value returned from the gp.predict() is 528 frames where each frame
consists of euler angles of 15 joints?

Also this 15 joints does not include the Hip joint, right?

Sincerely yours
Moon Jung

**Comment by Svito-zar** [comment]
Yes, that's correct those are the Euler angles of 15 joints. Not fully sure about Hip joint, but I don't think it was included since the model generated only the upper body movement

**Comment by moonryul** [comment]
Dear Svito-zar:

thanks for the answer. For the benefit of others who participate in this group, let me note that
your answer to my question is not right. Perhaps you have a different version of gesticulator in mind.

In demo.py code, we have:

    args = parse_args()
    
    audio_array, sample_rate = librosa.load(audio_file)
        
    feature_type, audio_dim = check_feature_type(args.model_file) 

    # 1. Load the model
    model = GesticulatorModel.load_from_checkpoint(  #MJ: model should be obtained in c# script once for all, not for every utterance
        args.model_file, inference_mode=True)
    # This interface is a wrapper around the model for predicting new gestures conveniently
    gp = GesturePredictor(model, feature_type)

    # 2. Predict the gestures with the loaded model
   
    predicted_motion = gp.predict_gestures(audio_text, audio_array, sample_rate )
    
    ==> I digged in the gp.predict_gestures() code, and have found that it returns the predicted motion
    in the form of theta*vec, which is the  representation of the exponential map.

    I wanted to return predicted_motion direclty to C# code of Unity, assuming that predicted_motion is in Euler angles. 
    
I found out what write_bvh() function converts the predicted motion in exponential map to Euler angles via
as follows:    
inv_data = data_pipeline.inverse_transform(anim_clip)

From: def write_bvh(datapipe_file, anim_clip, filename, fps):
    data_pipeline = joblib.load(datapipe_file[0])    
   
    inv_data = data_pipeline.inverse_transform(anim_clip) 

    writer = BVHWriter()
    for i in range(0, anim_clip.shape[0]): #MJ: anim_clip.shape = (1,520,45)
        with open(filename, "w") as f:            
            writer.write(inv_data[i], f, framerate=fps)

Sincerely,
